internal/text/text.go

```diff
@@ -0,0 +1,284 @@
+	Column   int
+	ID       string
+	Message  string
+}
+
+// FailureWriter is a writer that Failure.Println can accept.
+//
+// Both bytes.Buffer and bufio.Writer implement this.
+type FailureWriter interface {
+	WriteRune(rune) (int, error)
+	WriteString(string) (int, error)
+}
+
+// Fprintln prints the Failure to the writer with the given ordered fields.
+func (f *Failure) Fprintln(writer FailureWriter, fields ...FailureField) error {
+	if len(fields) == 0 {
+		fields = DefaultFailureFields
+	}
+	written := false
+	for i, field := range fields {
+		printColon := true
+		switch field {
+		case FailureFieldFilename:
+			filename := f.Filename
+			if filename == "" {
+				filename = "<input>"
+			}
+			if _, err := writer.WriteString(filename); err != nil {
+				return err
+			}
+			written = true
+		case FailureFieldLine:
+			line := strconv.Itoa(f.Line)
+			if line == "0" {
+				line = "1"
+			}
+			if _, err := writer.WriteString(line); err != nil {
+				return err
+			}
+			written = true
+		case FailureFieldColumn:
+			column := strconv.Itoa(f.Column)
+			if column == "0" {
+				column = "1"
+			}
+			if _, err := writer.WriteString(column); err != nil {
+				return err
+			}
+			written = true
+		case FailureFieldID:
+			if f.ID != "" {
+				if _, err := writer.WriteString(f.ID); err != nil {
+					return err
+				}
+				written = true
+			} else {
+				printColon = false
+			}
+		case FailureFieldMessage:
+			if f.Message != "" {
+				if _, err := writer.WriteString(f.Message); err != nil {
+					return err
+				}
+				written = true
+			} else {
+				printColon = false
+			}
+		default:
+			return fmt.Errorf("unknown FailureField: %v", field)
+		}
+		if printColon && i != len(fields)-1 {
+			if _, err := writer.WriteRune(':'); err != nil {
+				return err
+			}
+			written = true
+		}
+	}
+	if written {
+		_, err := writer.WriteRune('\n')
+		return err
+	}
+	return nil
+}
+
+// String implements fmt.Stringer.
+func (f *Failure) String() string {
+	filename := f.Filename
+	if filename == "" {
+		filename = "<input>"
+	}
+	line := strconv.Itoa(f.Line)
+	if line == "0" {
+		line = "1"
+	}
+	column := strconv.Itoa(f.Column)
+	if column == "0" {
+		column = "1"
+	}
+	buffer := bytes.NewBuffer(nil)
+	buffer.WriteString(filename)
+	buffer.WriteString(":")
+	buffer.WriteString(line)
+	buffer.WriteString(":")
+	buffer.WriteString(column)
+	buffer.WriteString(":")
+	if f.ID != "" {
+		buffer.WriteString(f.ID)
+		buffer.WriteString(" ")
+	}
+	buffer.WriteString(f.Message)
+	return buffer.String()
+}
+
+// NewFailuref is a helper that returns a new Failure.
+func NewFailuref(position scanner.Position, id string, format string, args ...interface{}) *Failure {
+	return &Failure{
+		ID:       id,
+		Filename: position.Filename,
+		Line:     position.Line,
+		Column:   position.Column,
+		Message:  fmt.Sprintf(format, args...),
+	}
+}
+
+// SortFailures sorts the Failures, by filename, line, column, id, message.
+func SortFailures(failures []*Failure) {
```

Comment:
Should we remove this function and move the `sort.Stable` call to `ParseColonSeparatedFailureFields`? 

This way none of the other packages need to explicitly call `SortFailures` and they can simply rely on it already being done. I don't see a reason where having the failures sorted would be a bad thing, so this seems useful overall.

Thoughts?

Reply:
Ah my mistake!